src/components/FocusBoundary.js

```diff
@@ -0,0 +1,62 @@
+import React from 'react';
+
+function isChildOf(target, child) {
+  if (!child) {
+    return false;
+  }
+
+  return target === child.parentElement
+    ? true
+    : isChildOf(target, child.parentElement);
+}
+
+export const FocusBoundary = ({ children, onChange, onEnter, onExit }) => {
+  const [ref, setRef] = React.useState();
+  const boundaryContainsFocus = React.useRef();
```

Comment:
Meanwhile, this is a ref because we _don't_ want to trigger a re-render or any other effects when this changes.